packages/cadl-ranch-specs/http/models/inheritance/main.cadl

```diff
@@ -1,60 +1,112 @@
-  discriminatorProperty: "A";
+@doc("The second level model in three levels inheritance with discriminator.")
+model Salmon extends Fish {
+  kind: "salmon";
+  friends?: Fish[];
+  hate?: Record<Fish>;
+  partner?: Fish;
 }
 
-model DerivedFromBaseClassWithDiscriminatorB extends BaseClassWithDiscriminator {
-  discriminatorProperty: "B";
+@doc("The third level model A in the three levels inheritance with discriminators.")
+model SawShark extends Shark {
+  sharktype: "saw";
 }
 
-@doc("Illustrates case where a basic model has polymorphic properties.")
-model ModelWithPolymorphicProperty {
-  @doc("Example polymorphic type property.")
-  polymorphicProperty: BaseClassWithDiscriminator;
+@doc("The third level model B in the non-discriminator three levels inheritance with discriminators.")
+model GoblinShark extends Shark {
+  sharktype: "goblin";
 }
 
 @scenario
-@scenarioDoc("Generate, send, and receive round-trip inherited model.")
-@route("/model")
+@scenarioDoc("Generate and send model. The valid input value is {name: \"abc\", age: 32, smart: true}")
+@route("/valid")
+@post
+op sendInheritanceValid(@body input: Siamese): OkResponse | ErrorResponse;
+
+@scenario
+@scenarioDoc("Generate and receive model. The return value is {name: \"abc\", age: 32, smart: true}")
+@route("/valid")
+@get
+op getInheritanceValid(): Siamese | ErrorResponse;
+
+@scenario
+@scenarioDoc("Generate, send, and receive round-trip bottom model.")
+@route("/valid")
 @put
-op setValue(@body input: BaseClass): BaseClass | ErrorResponse;
+op setInheritanceValid(@body input: Siamese): Siamese | ErrorResponse;
 
 @scenario
-@scenarioDoc("Generate, send, and receive round-trip model with a polymorphic property.")
-@route("/property")
+@route("/invalid")
+@scenarioDoc("Get a basic model that is invalid for the local strong kind.")
+@get
+op getInheritanceInvalid(): Pet | ErrorResponse;
+
+@scenario
+@scenarioDoc("Generate, send, and receive round-trip model that optional property is not presented.")
+@route("/empty")
 @put
-op setValueWithPolymorphicProperty(
-  @body input: ModelWithPolymorphicProperty
-): ModelWithPolymorphicProperty | ErrorResponse;
+op setInheritanceEmpty(@body input: Siamese): Pet | ErrorResponse;
+
+@scenario
+@scenarioDoc("Get a basic model that optional property is sent as null.")
+@route("/null")
+@get
+op getInheritanceWithNullProperty(): Pet | ErrorResponse;
+
+@scenario
+@route("/polymorphism/valid")
+@scenarioDoc("Generate, send, and receive round-trip models in three levels inheritance with 2 discriminators. The valid input is  { age: 1, kind: \"shark\",  sharktype: \"goblin\"}")
+@put
+op setPolymorphism(@body input: Fish): Fish | ErrorResponse;
+
+@scenario
+@route("/polymorphism/recursive")
+@scenarioDoc("Generate, send, and receive round-trip models has collection and dictionary properties referring to models in three levels inheritance with 2 discriminator")
+@put
+op setPolymorphismRecursively(@body input: Salmon): Salmon | ErrorResponse;
```

Comment:
I think for polymorphic models we always use the base class, is that correct @timotheeguerin

Reply:
Originally I use `Salmon` because the mockapi input/output an Salmon. But yes, base class `Fish` make more sense as an operation body.
Have changed to `Fish`.